src/peft/optimizers/riemannian.py

```diff
@@ -0,0 +1,291 @@
+        "lora_params": {},
+        "other_params": {},
+        "embedding": {}
+    }
+
+
+    for name, param in model.named_parameters():
+        if not param.requires_grad:
+            continue
+        # print(name, param.shape)
+        module = attrgetter(name)(model)
+        if isinstance(module, Embedding):
+            param_groups["embedding"][name] = param
+        elif "lora" in name:
+            param_groups["lora_params"][name] = param
+        else:
+            param_groups["other_params"][name] = param
+
+
+    lr = optimizer_kwargs["lr"]
+    weight_decay = optimizer_kwargs.get("weight_decay", 0.0)
+
+    optimizer_grouped_parameters = [
+        {
+            "params": list(param_groups["lora_params"].values()),
+            "weight_decay": weight_decay,
+            "lr": lr,
+            'is_lora': True
+        },
+        {
+            "params": list(param_groups["embedding"].values()),
+            "weight_decay": weight_decay,
+            "lr": lr_embedding,
+            'is_lora': False
+        },
+        {
+            "params": list(param_groups["other_params"].values()),
+            "weight_decay": weight_decay,
+            "lr": lr,
+            'is_lora': False
+        }
+    ]
+
+    optimizer_kwargs.update({'reg':reg})
+    optimizer = riemannian_AdamW(optimizer_grouped_parameters, **optimizer_kwargs)
+    if optimizer_cls.__name__ == "Adam8bit":
+        raise Exception('bitsandbytes not supported yet')
+    
+    return optimizer
+
+
+class riemannian_AdamW(Optimizer):
+    """
+    Adapt from AdamW code in transformers.optimization.AdamW
+
+    Parameters:
+        params (`Iterable[nn.parameter.Parameter]`):
+            Iterable of parameters to optimize or dictionaries defining parameter groups.
+        lr (`float`, *optional*, defaults to 0.001):
+            The learning rate to use.
+        betas (`Tuple[float,float]`, *optional*, defaults to `(0.9, 0.999)`):
+            Adam's betas parameters (b1, b2).
+        eps (`float`, *optional*, defaults to 1e-06):
+            Adam's epsilon for numerical stability.
+        weight_decay (`float`, *optional*, defaults to 0.0):
+            Decoupled weight decay to apply.
+        correct_bias (`bool`, *optional*, defaults to `True`):
+            Whether or not to correct bias in Adam (for instance, in Bert TF repository they use `False`).
+        no_deprecation_warning (`bool`, *optional*, defaults to `False`):
+            A flag used to disable the deprecation warning (set to `True` to disable the warning).
+    """
+
+    def __init__(
+        self,
+        params: Iterable[nn.parameter.Parameter],
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-6,
+        weight_decay: float = 0.0,
+        correct_bias: bool = True,
+        no_deprecation_warning: bool = False,
+        reg: float = 1e-6
+    ):
+        print('CREATE Riemannian AdamW')
+        if not no_deprecation_warning:
```

Comment:
I follow HF transformer's implementation https://github.com/huggingface/transformers/blob/v4.42.0/src/transformers/optimization.py#L558. I guess it's because the group might want to further modify the implementation. Anyway, I'll just remove the warning.